### PR TITLE
Make phpType definitions in generated DAO objects more accurate.

### PR DESCRIPTION
**Pull request by braders**
Overview
----------------------------------------
Make phpType definitions in generated DAO objects more accurate.

DAO objects are generated from the schema XML files, and list every database field as a property. However, there are some inaccuracies with how these properties are documented, as described below. This PR changes the code generator, so that the properties in DAO classes have more accurate PHPDoc comments.

Before
----------------------------------------
Take the follow examples from `CRM_ACL_DAO_ACL`: 
```
/**
 * Unique table ID
 *
 * @var int
 */
public $id;

```

There are a couple of issues with the propery comments that are currently generated:

1. The property may be null, in the above example because `$id` is the primary key, and so may be null when a new object is being created (the ID is only set on save).
2. For other properties the database field may not be required, in which case `null` is also a valid value.
3. As far as I can tell, database values are always read from the database as a `string`. It would be good if someone else could validate what I'm seeing here, but reading and then `var_dump`'ing a DAO object always shows the type as string. This seems to be true regardless of the SQL value is `int`, `bool`, `timestamp` etc.
4. Currently the PHPDoc annotations in DAO objects contain types such as `timestamp`. These are not valid PHP types, and therefore should be `string`.

After
----------------------------------------
The above example would become:

```
  /**
   * Unique table ID
   *
   * @var int|string|null
   *   (SQL type: int unsigned)
   *   Note that values will be retrieved from the database as a string.
   */
  public $id;
```

A few changes have been made:

1. Type `string` is added to all properties, as the data will be read out of the database in this format
2. The `null` value is documented where appropriate (non-required and primary-key fields)
3. The SQL type is added as a comment
4. A reminder about values being read as strings is added

Technical Details
----------------------------------------
I suspect the fact that values always seem to be read out of the database as a string will come as a surprise for a lot of people - I certainly don't find it intuitive, and I suspect it could be the cause of bugs. As I said above, it would be good for someone to sense-check this, and confirm there are not cases where this is not the case.

Longer term the DAO objects could transform the values into more appropriate types when they are read from the database. However, this is a bigger change that could cause breakage - for example, there may be code that relies on a boolean being returned as '0' or '1' rather than an actual boolean. For now this change at least makes the confusion clear, even if it doesn't do anything to resolve it!

Comments
----------------------------------------
I've not committed the 100+ re-generated DAO files, in order to make it possible to effectively review the code which has actaully changed. However, I have re-generated the files locally, and can share or commit as appropriate.

### CODE REVIEW

File: xml/templates/dao.tpl [hunk 1]
```diff
@@ -72,7 +72,11 @@ class {$table.className} extends CRM_Core_DAO {ldelim}
      * {$field.comment|regex_replace:"/\n[ ]*/":"\n* "}
      *
 {/if}
-     * @var {$field.phpType}
+     * @var {$field.phpType}{if $field.phpNullable}|null
+{/if}
+
```
Comment (braders) [inline xml/templates/dao.tpl:77]: I had some issues with smarty ignoring my newline after the closing {/if}, hence the slightly odd formatting here.

**Comment by colemanw** [comment]
Thanks for taking this on @braders. Some background that might help:
- The DAO object is pretty minimalist, and a lot of hacks and workarounds have build up over the years to deal with its inadequacies. Yep, it returns the wrong data type, fails to unserialize array data, can't handle joins, doesn't understand option lists, can't deal with NULL values.
- Fortunately we now have APIv4, which *can* do all of the above, and is the recommended way to interact with the database, for all but the lowest-level code.
- I see you've had an encounter with the ironically ugly `PHP_Beautifier` class which messes up more than it fixes. AFAIK it is only used once in our codebase (generating DAO files). I'd love to get rid of it, simply by having the `dao.tpl` file use correct formatting in the first place so it doesn't need any beautification. Maybe that's something you'd like to lend a hand with?

**Comment by braders** [comment]
@colemanw Thanks for the feedback. I had no ideal about `PHP_Beautifier`, so good to know the buginess I was seeing isn't the fault of `Smarty`. 

Locally, I've just commented out the lines which actually run `PHP_Beautifier` and the resuling DAO files are largely the same, but with the indentation all over the place. There are also some differences in whether short or long array syntax is used, but that doesn't feel like such a big deal.

I suspect a multi-step approach to removing `PHP_Beautifier` would be the way to go:

1. Tidy up dao.tpl to use sensible indentation. Currently its all over the place.
2. Remove reference to `PHP_Beautifier` in `CRM_Core_CodeGen_Util_Template`.
3. Remove `PHP_Beautifier` from the Civi packages.

Note that I've only been looking at `PHP_Beautifier`  in the context of DAO files. It also seems to be used to generate SQL files at the moment.

**Comment by colemanw** [comment]
Yea I think the only real blocker to generating DAOs without `PHP_Beautifier` is the ability to print arrays nicely (with correct indentation, square brackets, and omitting the `0 =>`, `1 =>`, etc for non-associative arrays).

But `PHP_Beautifier` isn't taking care of that stuff anyway, I think it just does indentation and then we have to fix the rest up here:
https://github.com/civicrm/civicrm-core/blob/2be680289f9a8d3971807b8b4000d5405852e856/CRM/Core/CodeGen/Util/Template.php#L86-L89

**Comment by braders** [comment]
@colemanw I've had a play with tidying up the dao.tpl template here: https://github.com/braders/civicrm-core/commit/3bbb5cbb5f48ac489c642e06df04990ab4eadf56.

If I comment out the PHP_Beutifier lines in `CRM_Core_CodeGen_Util_Template`, specifically `$contents = $this->beautifier->get();` you can see how far this gets us. There are some issues with indentation and arrays which I'm not sure how we'd resolve without `PHP_Beautifier` unfortunately. It's the sort of thing where it might be best for you to grab that template file and look for yourself!

**Comment by colemanw** [comment]
I don't want scope-creep to stall this PR, so to be clear the discussion about `PHP_Beautifier` is for a followup project.
In reviewing this one, I regenerated the DAO files locally. For some reason I encountered a ton of warnings in the console when I did so:
```
Warning: Trying to access array offset on value of type null in /civicrm/packages/PHP/Beautifier.php on line 1306
```
That's not necessarily a blocker if we're planning to remove that tool anyway, but it's curious. I don't think I saw that before this PR.

My only other review comment is that the repetition of the line `Note that values will be retrieved from the database as a string.` feels like too much boilerplate. I think it would be sufficient to add that note once somewhere and not repeat it hundreds of times.

**Comment by braders** [comment]
@colemanw I'm not seeing any warnings about array access, but if I step through the Beautifier code I can see that it is trying to read an array offset on `null`. What version of PHP are you using? Curiously running `echo null[0];` in an interactive terminal throws no warnings for me in the 7.x stream, but I wonder if this has changed in PHP 8. 

I can see why you say that the repeated comment is too much repitition, but I'm not sure I agree. Most developers are unlikely to be reading the DAO classes in their entirety, and instead will be referencing the bits which are relevant to them. By putting the comment in each propery description it means the text will be exposed through IDE's autocomplete at the point the properties are being used - which is at the point when the message is valuable.

**Comment by colemanw** [comment]
Ok fair enough @braders. I have regenerated DAO files for core + core extensions, and pushed to this branch. It's very merge-conflict-prone so giving it merge-on-pass label.

**Comment by colemanw** [comment]
I think you could followup with a PR for braders@3bbb5cb as it doesn't harm anything & cleans up the whitespace to move us closer to ending `PHP_Beautifier`'s reign of terror.

**Comment by seamuslee001** [comment]
I just rebased as another PR that got merged caused a conflict on the Case DAO

**Comment by colemanw** [comment]
@braders this is merged, want to do that followup PR for the tpl cleanup?

**Pull request merged**